--- uboot-2024.04/drivers/mtd/nand/raw/mxs_nand_dt.c	2024-04-02 22:58:58.000000000 
+++ uboot-2024.04/drivers/mtd/nand/raw/mxs_nand_dt.c	2026-06-22 10:24:28.000000000 
@@ -25,6 +25,16 @@
 	int max_chain_delay; /* See the async EDO mode */
 };
 
+static const struct mxs_nand_dt_data mxs_nand_imx23_data = {
+	.max_ecc_strength_supported = 20,
+	.max_chain_delay = 16000,
+};
+
+static const struct mxs_nand_dt_data mxs_nand_imx28_data = {
+	.max_ecc_strength_supported = 20,
+	.max_chain_delay = 16000,
+};
+
 static const struct mxs_nand_dt_data mxs_nand_imx6q_data = {
 	.max_ecc_strength_supported = 40,
 	.max_chain_delay = 12000,
@@ -46,6 +56,14 @@
 };
 
 static const struct udevice_id mxs_nand_dt_ids[] = {
+	{
+		.compatible = "fsl,imx23-gpmi-nand",
+		.data = (unsigned long)&mxs_nand_imx23_data,
+	},
+	{
+		.compatible = "fsl,imx28-gpmi-nand",
+		.data = (unsigned long)&mxs_nand_imx28_data,
+	},
 	{
 		.compatible = "fsl,imx6q-gpmi-nand",
 		.data = (unsigned long)&mxs_nand_imx6q_data,
