--- a/include/configs/mx23_olinuxino.h	2024-04-02 22:58:58.000000000 
+++ a/include/configs/mx23_olinuxino.h	2026-07-05 21:48:30.000000000 
@@ -6,13 +6,13 @@
 #define __CONFIGS_MX23_OLINUXINO_H__
 
 /* U-Boot Commands */
 
 /* Memory configuration */
 #define PHYS_SDRAM_1			0x40000000	/* Base address */
-#define PHYS_SDRAM_1_SIZE		0x08000000	/* Max 128 MB RAM */
+#define PHYS_SDRAM_1_SIZE		0x04000000	/* Max 64 MB RAM */
 #define CFG_SYS_SDRAM_BASE		PHYS_SDRAM_1
 
 /* Status LED */
 
 /* USB */
 
