--- a/drivers/gpio/gpio-74x164.c	2026-04-30 21:17:22.000000000 
+++ a/drivers/gpio/gpio-74x164.c	2026-06-22 12:37:24.000000000 
@@ -21,6 +21,7 @@
 	struct gpio_chip	gpio_chip;
 	struct mutex		lock;
 	struct gpio_desc	*gpiod_oe;
+	struct gpio_desc	*gpiod_strobe;
 	u32			registers;
 	/*
 	 * Since the registers are chained, every byte sent will make
@@ -60,12 +61,26 @@
 	u8 pin = offset % 8;
 
 	mutex_lock(&chip->lock);
+	if (chip->gpiod_strobe) {
+		if (chip->gpiod_oe)
+			gpiod_set_value_cansleep(chip->gpiod_oe, 0);
+	}
+
 	if (val)
 		chip->buffer[bank] |= (1 << pin);
 	else
 		chip->buffer[bank] &= ~(1 << pin);
 
 	__gen_74x164_write_config(chip);
+
+	/*  Latch data to output pins*/
+	if (chip->gpiod_strobe) {
+		gpiod_set_value_cansleep(chip->gpiod_strobe, 0);
+		gpiod_set_value_cansleep(chip->gpiod_strobe, 1);
+
+		if (chip->gpiod_oe)
+			gpiod_set_value_cansleep(chip->gpiod_oe, 1);
+	}
 	mutex_unlock(&chip->lock);
 }
 
@@ -79,6 +94,11 @@
 	unsigned long bitmask;
 
 	mutex_lock(&chip->lock);
+	if (chip->gpiod_strobe) {
+		if (chip->gpiod_oe)
+			gpiod_set_value_cansleep(chip->gpiod_oe, 0);
+	}
+
 	for_each_set_clump8(offset, bankmask, mask, chip->registers * 8) {
 		bank = chip->registers - 1 - offset / 8;
 		bitmask = bitmap_get_value8(bits, offset) & bankmask;
@@ -87,6 +107,15 @@
 		chip->buffer[bank] |= bitmask;
 	}
 	__gen_74x164_write_config(chip);
+
+	/*  Latch data to output pins*/
+	if (chip->gpiod_strobe) {
+		gpiod_set_value_cansleep(chip->gpiod_strobe, 0);
+		gpiod_set_value_cansleep(chip->gpiod_strobe, 1);
+
+		if (chip->gpiod_oe)
+			gpiod_set_value_cansleep(chip->gpiod_oe, 1);
+	}
 	mutex_unlock(&chip->lock);
 }
 
@@ -127,6 +156,11 @@
 	if (IS_ERR(chip->gpiod_oe))
 		return PTR_ERR(chip->gpiod_oe);
 
+	chip->gpiod_strobe = devm_gpiod_get_optional(&spi->dev, "strobe",
+			GPIOD_OUT_LOW);
+	if (IS_ERR(chip->gpiod_strobe))
+		return PTR_ERR(chip->gpiod_strobe);
+
 	spi_set_drvdata(spi, chip);
 
 	chip->gpio_chip.label = spi->modalias;
@@ -145,13 +179,25 @@
 
 	mutex_init(&chip->lock);
 
+	if (chip->gpiod_strobe) {
+		gpiod_set_value_cansleep(chip->gpiod_strobe, 1);
+	}
+
 	ret = __gen_74x164_write_config(chip);
 	if (ret) {
 		dev_err(&spi->dev, "Failed writing: %d\n", ret);
 		goto exit_destroy;
 	}
 
-	gpiod_set_value_cansleep(chip->gpiod_oe, 1);
+	/*  Latch data to output pins*/
+	if (chip->gpiod_strobe) {
+		gpiod_set_value_cansleep(chip->gpiod_strobe, 0);
+		gpiod_set_value_cansleep(chip->gpiod_strobe, 1);
+	}
+
+	if (chip->gpiod_oe) {
+		gpiod_set_value_cansleep(chip->gpiod_oe, 1);
+	}
 
 	ret = gpiochip_add_data(&chip->gpio_chip, chip);
 	if (!ret)
@@ -167,7 +213,10 @@
 {
 	struct gen_74x164_chip *chip = spi_get_drvdata(spi);
 
-	gpiod_set_value_cansleep(chip->gpiod_oe, 0);
+	if (chip->gpiod_oe) {
+		gpiod_set_value_cansleep(chip->gpiod_oe, 0);
+	}
+
 	gpiochip_remove(&chip->gpio_chip);
 	mutex_destroy(&chip->lock);
 }
@@ -175,6 +224,8 @@
 static const struct spi_device_id gen_74x164_spi_ids[] = {
 	{ .name = "74hc595" },
 	{ .name = "74lvc594" },
+	{ .name = "cd54hc4094" },
+	{ .name = "cd74hc4094" },
 	{},
 };
 MODULE_DEVICE_TABLE(spi, gen_74x164_spi_ids);
@@ -182,6 +233,8 @@
 static const struct of_device_id gen_74x164_dt_ids[] = {
 	{ .compatible = "fairchild,74hc595" },
 	{ .compatible = "nxp,74lvc594" },
+	{ .compatible = "ti,cd54hc4094" },
+	{ .compatible = "ti,cd74hc4094" },
 	{},
 };
 MODULE_DEVICE_TABLE(of, gen_74x164_dt_ids);
--- a/Documentation/devicetree/bindings/gpio/fairchild,74hc595.yaml	2026-04-30 21:17:22.000000000 
+++ a/Documentation/devicetree/bindings/gpio/fairchild,74hc595.yaml	2026-06-22 12:13:57.000000000 
@@ -14,6 +14,8 @@
     enum:
       - fairchild,74hc595
       - nxp,74lvc594
+      - ti,cd54hc4094
+      - ti,cd74hc4094
 
   reg:
     maxItems: 1
@@ -31,6 +33,10 @@
 
   enable-gpios:
     description: GPIO connected to the OE (Output Enable) pin.
+    maxItems: 1
+
+  strobe-gpios:
+    description: GPIO connected to the STROBE pin
     maxItems: 1
 
 patternProperties:
